### PR TITLE
Enable router timeout configuration and parameters object

### LINKED ISSUE #991: Config Robustness
The runtime exposes a couple of setters to change different configurations that different runtime views rely on, therefore the views should only be initialized after the configurations have been set. If they interleave, then some configurations can be ignored. It might be a good idea to force an order.

### LINKED ISSUE #995: RPC Timeout
Make this value configurable from the runtime.

### PR DESCRIPTION
**Pull request by no2chem**
## Overview

Description: This PR moves all settings of the runtime into the CorfuRuntimeParameters object, and allows RPC timeouts to be set at construction time if the user generates a runtime with a CorfuRuntimeParameters object. Previously, a NettyClientRouter would be created as soon as the runtime connected, inheriting the defaults rather than the user-defined settings.

Why should this be merged: This PR enables setting of timeouts on the router, which allows users to reconfigure for a slow network. It also prevents race conditions where the runtime is active while settings are being changed.

Related issue(s) (if applicable): Fixes #995, Fixes #991, Closes #1027 

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/runtime/exceptions/NetworkException.java [hunk 1]
```diff
@@ -1,22 +1,24 @@
...
+import org.corfudb.util.NodeLocator;
 
 /**
  * Created by mwei on 12/14/15.
  */
 public class NetworkException extends RuntimeException {
 
     @Getter
-    String endpoint;
+    NodeLocator node;
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/exceptions/NetworkException.java]: CRITICAL Make "node" transient or serializable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1948)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 2]
```diff
@@ -659,12 +613,197 @@ public synchronized CorfuRuntime connect() {
...
+        parameters.tlsEnabled = true;
+        return this;
+    }
+
+    /**
+     * Enable SASL Plain Text.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     **/
+    @Deprecated
+    public CorfuRuntime enableSaslPlainText(String usernameFile, String passwordFile) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 3]
```diff
@@ -659,12 +613,197 @@ public synchronized CorfuRuntime connect() {
...
+        this(CorfuRuntimeParameters.builder().build());
+    }
+
+    /**
+     * Parse a configuration string and get a CorfuRuntime.
+     *
+     * @param configurationString The configuration string to parse.
+     */
+    @Deprecated
+    public CorfuRuntime(String configurationString) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 4]
```diff
@@ -659,12 +613,197 @@ public synchronized CorfuRuntime connect() {
...
+        return this;
+    }
+
+    /** Set the timeout of the fast loader, in minutes.
+     *
+     * @param timeout   The number of minutes to wait.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     */
+    @Deprecated
+    public CorfuRuntime setTimeoutInMinutesForFastLoading(int timeout) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 5]
```diff
@@ -659,12 +613,197 @@ public synchronized CorfuRuntime connect() {
...
+        return this;
+    }
+
+    /** Set the number of cache entries.
+     *
+     * @param numCacheEntries   The number of cache entries.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     */
+    @Deprecated
+    public CorfuRuntime setNumCacheEntries(long numCacheEntries) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 6]
```diff
@@ -659,12 +613,197 @@ public synchronized CorfuRuntime connect() {
...
+    }
+
+
+    /** Set the write retry time.
+     *
+     * @param writeRetry   The number of times to retry writes.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     */
+    @Deprecated
+    public CorfuRuntime setWriteRetry(int writeRetry) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 7]
```diff
@@ -659,12 +613,197 @@ public synchronized CorfuRuntime connect() {
...
+
+    /**
+     * Whether or not to use the fast loader.
+     *
+     * @param enable True, if the fast loader should be used, false otherwise.
+     * @return A CorfuRuntime to support chaining.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     */
+    @Deprecated
+    public CorfuRuntime setLoadSmrMapsAtConnect(boolean enable) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 8]
```diff
@@ -659,12 +613,197 @@ public synchronized CorfuRuntime connect() {
...
+    // endregion
+
+    // region Deprecated Setters
+
+    /**
+     * Enable TLS.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     **/
+    @Deprecated
+    public CorfuRuntime enableTls(String keyStore, String ksPasswordFile, String trustStore,
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 9]
```diff
@@ -659,12 +613,197 @@ public synchronized CorfuRuntime connect() {
...
+
+    /**
+     * Whether or not to disable backpointers
+     *
+     * @param disable True, if the cache should be disabled, false otherwise.
+     * @return A CorfuRuntime to support chaining.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     */
+    @Deprecated
+    public CorfuRuntime setBackpointersDisabled(boolean disable) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 10]
```diff
@@ -659,12 +613,197 @@ public synchronized CorfuRuntime connect() {
...
+        return this;
+    }
+
+    /** Set the cache expiration time.
+     *
+     * @param expiryTime   The time before cache expiration, in seconds.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     */
+    @Deprecated
+    public CorfuRuntime setCacheExpiryTime(int expiryTime) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 11]
```diff
@@ -659,12 +613,197 @@ public synchronized CorfuRuntime connect() {
...
+        return this;
+    }
+
+    /** Set the trim retry time.
+     *
+     * @param trimRetry   The number of times to retry on trims.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     */
+    @Deprecated
+    public CorfuRuntime setTrimRetry(int trimRetry) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 12]
```diff
@@ -659,12 +613,197 @@ public synchronized CorfuRuntime connect() {
...
+
+    // Below are deprecated methods which should no longer used and may be deprecated in the future.
+
+    // region Deprecated Constructors
+    /**
+     * Constructor for CorfuRuntime.
+     *
+     **/
+    @Deprecated
+    public CorfuRuntime() {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 13]
```diff
@@ -659,12 +613,197 @@ public synchronized CorfuRuntime connect() {
...
+
+    /**
+     * Whether or not hole filling is disabled
+     *
+     * @param disable True, if hole filling should be disabled
+     * @return A CorfuRuntime to support chaining.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     */
+    @Deprecated
+    public CorfuRuntime setHoleFillingDisabled(boolean disable) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 14]
```diff
@@ -659,12 +613,197 @@ public synchronized CorfuRuntime connect() {
...
+
+    /**
+     * Whether or not to disable the cache
+     *
+     * @param disable True, if the cache should be disabled, false otherwise.
+     * @return A CorfuRuntime to support chaining.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     */
+    @Deprecated
+    public CorfuRuntime setCacheDisabled(boolean disable) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 15]
```diff
@@ -659,12 +613,197 @@ public synchronized CorfuRuntime connect() {
...
+        return this;
+    }
+
+    /** Set the bulk read size.
+     *
+     * @param size  The bulk read size.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     */
+    @Deprecated
+    public CorfuRuntime setBulkReadSize(int size) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java [hunk 16]
```diff
@@ -206,46 +205,52 @@ public NettyClientRouter(String endpoint) {
      * @param host Host to connect to.
      * @param port Port to connect to.
      */
+    @Deprecated
     public NettyClientRouter(String host, Integer port) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java [hunk 17]
```diff
@@ -165,37 +165,36 @@ public void setEpoch(long epoch){
...
+    private final CorfuRuntimeParameters parameters;
+
+    @Deprecated
+    @Override
+    public Integer getPort() {
+        return node.getPort();
+    }
+
+    @Deprecated
+    public String getHost() {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java [hunk 18]
```diff
@@ -165,37 +165,36 @@ public void setEpoch(long epoch){
...
-
-    private String saslPlainTextPasswordFile;
-
     /**
      * Creates a new NettyClientRouter connected to the specified endpoint.
      *
      * @param endpoint Endpoint to connect to.
      */
+    @Deprecated
     public NettyClientRouter(String endpoint) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java [hunk 19]
```diff
@@ -206,46 +205,52 @@ public NettyClientRouter(String endpoint) {
...
     }
 
+    /**
+     * Creates a new NettyClientRouter connected to the specified host and port.
+     *
+     * @param host Host to connect to.
+     * @param port Port to connect to.
+     */
+    @Deprecated
     public NettyClientRouter(String host, Integer port, Boolean tls,
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java]: MAJOR Add the missing @deprecated Javadoc tag. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AMissingDeprecatedCheck)
INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

**Comment by corfu-repo** [comment]
SonarQube analysis reported 26 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 6 major
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 18 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [NettyClientRouter.java#L32](https://github.com/CorfuDB/CorfuDB/blob/0e4e5c597e6596d143e9603e1a0e661a98f4ddc3/runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java#L32): Remove this unused import 'java.util.UUID'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **0e4e5c5**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #1090 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/2qo4z0VsSLHImkhoqY22oPN2yYZ4nXZD)

File: runtime/src/main/java/org/corfudb/util/NodeLocator.java [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+    final String host;
+
+    /** The port number on the host the node is located on. */
+    final int port;
+
+    /** The ID of the node. Can be null if node id matching is not requested. */
+    @Builder.Default private UUID nodeId = null;
+
+    /** A map of options. */
+    @Singular final Map<String, String> options;
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/util/NodeLocator.java]: CRITICAL Make "options" private or transient. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1948)

**Comment by corfu-repo** [comment]
SonarQube analysis reported 18 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 14 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [NettyClientRouter.java#L32](https://github.com/CorfuDB/CorfuDB/blob/d348ac24572774b9ba300fbae4cd740874a9c8e7/runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java#L32): Remove this unused import 'java.util.UUID'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 21]
```diff
@@ -52,32 +58,121 @@
...
+        /** True, if optimistic undo logging is disabled. */
+        @Default boolean optimisticUndoDisabled = false;
 
         /**
-         * True, if optimistic undo logging is disabled.
+         * Use fast loader to restore objects on connection.
+         *
+         * <p>If using this utility, you need to be sure that no one
+         * is accessing objects until the tables are loaded
+         * (i.e. when connect return)
```
Comment (annym) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: returnS

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 22]
```diff
@@ -52,32 +58,121 @@
...
+        @Default boolean useFastLoader = false;
+
+        /** Set the bulk read size. */
+        @Default int bulkReadSize = 10;
 
         /**
-         * Number of times to attempt to read before hole filling.
+         * How much time the Fast Loader has to get the maps up to date.
+         *
+         * <p>Once the timeout is reached, the Fast Loader give up. Every maps that are
```
Comment (annym) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: Some typos: give(S)... Every map() not up ...

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 23]
```diff
@@ -659,12 +613,197 @@ public synchronized CorfuRuntime connect() {
...
-                    .setBatchReadSize(getBulkReadSize())
-                    .setTimeoutInMinutesForLoading(timeoutInMinutesForFastLoading);
+                    .setBatchReadSize(parameters.getBulkReadSize())
+                    .setTimeoutInMinutesForLoading((int) parameters.fastLoaderTimeout.toMinutes());
             fastLoader.loadMaps();
         }
         return this;
     }
+
+    // Below are deprecated methods which should no longer used and may be deprecated in the future.
```
Comment (annym) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: typo: no longer (be) used..

**Comment by no2chem** [comment]
@annym Ok, great. Rebased and waiting for CI now.

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 24]
```diff
@@ -659,12 +613,199 @@ public synchronized CorfuRuntime connect() {
...
+
+    /**
+     * Whether or not to disable the cache
+     *
+     * @param disable True, if the cache should be disabled, false otherwise.
+     * @return A CorfuRuntime to support chaining.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     */
+    @Deprecated
+    public CorfuRuntime setCacheDisabled(boolean disable) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 25]
```diff
@@ -659,12 +613,199 @@ public synchronized CorfuRuntime connect() {
...
+        return this;
+    }
+
+    /** Set the bulk read size.
+     *
+     * @param size  The bulk read size.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     */
+    @Deprecated
+    public CorfuRuntime setBulkReadSize(int size) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 26]
```diff
@@ -659,12 +613,199 @@ public synchronized CorfuRuntime connect() {
...
+        parameters.tlsEnabled = true;
+        return this;
+    }
+
+    /**
+     * Enable SASL Plain Text.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     **/
+    @Deprecated
+    public CorfuRuntime enableSaslPlainText(String usernameFile, String passwordFile) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 27]
```diff
@@ -659,12 +613,199 @@ public synchronized CorfuRuntime connect() {
...
+    }
+
+    /**
+     * Parse a configuration string and get a CorfuRuntime.
+     *
+     * @param configurationString The configuration string to parse.
+     * @deprecated Use {@link CorfuRuntime#fromParameters(CorfuRuntimeParameters)}
+     */
+    @Deprecated
+    public CorfuRuntime(String configurationString) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 28]
```diff
@@ -659,12 +613,199 @@ public synchronized CorfuRuntime connect() {
...
+        return this;
+    }
+
+    /** Set the timeout of the fast loader, in minutes.
+     *
+     * @param timeout   The number of minutes to wait.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     */
+    @Deprecated
+    public CorfuRuntime setTimeoutInMinutesForFastLoading(int timeout) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 29]
```diff
@@ -659,12 +613,199 @@ public synchronized CorfuRuntime connect() {
...
+        return this;
+    }
+
+    /** Set the number of cache entries.
+     *
+     * @param numCacheEntries   The number of cache entries.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     */
+    @Deprecated
+    public CorfuRuntime setNumCacheEntries(long numCacheEntries) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 30]
```diff
@@ -659,12 +613,199 @@ public synchronized CorfuRuntime connect() {
...
+    }
+
+
+    /** Set the write retry time.
+     *
+     * @param writeRetry   The number of times to retry writes.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     */
+    @Deprecated
+    public CorfuRuntime setWriteRetry(int writeRetry) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 31]
```diff
@@ -659,12 +613,199 @@ public synchronized CorfuRuntime connect() {
...
+
+    /**
+     * Whether or not to use the fast loader.
+     *
+     * @param enable True, if the fast loader should be used, false otherwise.
+     * @return A CorfuRuntime to support chaining.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     */
+    @Deprecated
+    public CorfuRuntime setLoadSmrMapsAtConnect(boolean enable) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 32]
```diff
@@ -659,12 +613,199 @@ public synchronized CorfuRuntime connect() {
...
+    // endregion
+
+    // region Deprecated Setters
+
+    /**
+     * Enable TLS.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     **/
+    @Deprecated
+    public CorfuRuntime enableTls(String keyStore, String ksPasswordFile, String trustStore,
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 33]
```diff
@@ -659,12 +613,199 @@ public synchronized CorfuRuntime connect() {
...
+
+    /**
+     * Whether or not to disable backpointers
+     *
+     * @param disable True, if the cache should be disabled, false otherwise.
+     * @return A CorfuRuntime to support chaining.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     */
+    @Deprecated
+    public CorfuRuntime setBackpointersDisabled(boolean disable) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 34]
```diff
@@ -659,12 +613,199 @@ public synchronized CorfuRuntime connect() {
...
+        return this;
+    }
+
+    /** Set the cache expiration time.
+     *
+     * @param expiryTime   The time before cache expiration, in seconds.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     */
+    @Deprecated
+    public CorfuRuntime setCacheExpiryTime(int expiryTime) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 35]
```diff
@@ -659,12 +613,199 @@ public synchronized CorfuRuntime connect() {
...
+        return this;
+    }
+
+    /** Set the trim retry time.
+     *
+     * @param trimRetry   The number of times to retry on trims.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     */
+    @Deprecated
+    public CorfuRuntime setTrimRetry(int trimRetry) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 36]
```diff
@@ -659,12 +613,199 @@ public synchronized CorfuRuntime connect() {
...
+    // Below are deprecated methods which should no longer be
+    // used and may be deprecated in the future.
+
+    // region Deprecated Constructors
+    /**
+     * Constructor for CorfuRuntime.
+     * @deprecated Use {@link CorfuRuntime#fromParameters(CorfuRuntimeParameters)}
+     **/
+    @Deprecated
+    public CorfuRuntime() {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

File: runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java [hunk 37]
```diff
@@ -659,12 +613,199 @@ public synchronized CorfuRuntime connect() {
...
+
+    /**
+     * Whether or not hole filling is disabled
+     *
+     * @param disable True, if hole filling should be disabled
+     * @return A CorfuRuntime to support chaining.
+     * @deprecated  Deprecated, set using {@link CorfuRuntimeParameters} instead.
+     */
+    @Deprecated
+    public CorfuRuntime setHoleFillingDisabled(boolean disable) {
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1133)

**Comment by corfu-repo** [comment]
SonarQube analysis reported 15 issues
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 14 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [NettyClientRouter.java#L32](https://github.com/CorfuDB/CorfuDB/blob/963842aff593702a8cd95e5300ca44ebdb30d9be/runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java#L32): Remove this unused import 'java.util.UUID'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**